### ISSUE TITLE
Rule proposal: no-unsafe-readonly-mutable-assignment

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#5954]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

From https://github.com/danielnixon/eslint-plugin-total-functions:

> Bans unsafe assignment of readonly values to mutable values (which can lead to surprising mutation in the readonly value). This includes passing readonly values as arguments to functions that expect mutable parameters.
>
> For examples of assignment that this rule considers valid and invalid, see [no-unsafe-readonly-mutable-assignment.test.ts](https://github.com/danielnixon/eslint-plugin-total-functions/blob/master/src/rules/no-unsafe-readonly-mutable-assignment.test.ts).
>
> See https://github.com/microsoft/TypeScript/issues/13347 for a request to fix this at the language level.

Though, I wonder if we can shorten the name a bit?

### Fail Cases

```typescript
type MutableA = { a: string };
type ReadonlyA = { readonly a: string };

const mutate = (mut: MutableA): void => {
    mut.a = "whoops";
};

const readonlyA: ReadonlyA = { a: "readonly?" };
mutate(readonlyA);
```

### Pass Cases

```typescript
type MutableA = { a: string };
type ReadonlyA = { readonly a: string };

const read = (mut: ReadonlyA): void => {
    // No longer allowed:
    // mut.a = "whoops";
};

const readonlyA: ReadonlyA = { a: "readonly?" };
read(readonlyA);
```

### Additional Info

Forking conversation out from https://github.com/danielnixon/eslint-plugin-total-functions/issues/665.

**Comment by danielnixon** [comment]
This may be a better approach: https://github.com/danielnixon/eslint-plugin-total-functions/issues/708

**Comment by bradzacher** [comment]
I believe that this will be fixed by TS 5.5 which is including a new flag `--enforceReadonly`

https://github.com/microsoft/TypeScript/pull/58296
